test/core/test_patterns.py

```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 """
-Unit tests for mathics pattern matching
+Unit tests for mathics.core.pattern
```

Comment:
Here is the part that feels a little weird to me. We are testing a lower-level function using higher-level tests.  I am a little bit sensitive to this kind of thing, because this less-good practice is pervasive in this code.

At one time, we didn't have pytests. There were doctests that took the place of pytests. 

The MatchQ tests to me should be put in `test.buitin.testing_expressions.expression_tests` and the rule tests put wherever rules are located. 

Tests here could (and should be) to build a ExpressionPattern and run `match` on it.

Reply:
> Here is the part that feels a little weird to me. We are testing a lower-level function using higher-level tests. I am a little bit sensitive to this kind of thing, because this less-good practice is pervasive in this code.
> 
> At one time, we didn't have pytests. There were doctests that took the place of pytests. It took us a while to get unburied from that practice.
> 
> The MatchQ tests to me should be put in `test.buitin.testing_expressions.expression_tests` and the rule tests put wherever rules are located.
> 
> Tests here could (and should be) to build a ExpressionPattern and run `match` on it.

Yes, but in any case, at some point we can rewrite these tests as lower-level tests. I do not think that these tests actually test the behavior of `MatchQ` or `ReplaceAll`, but the low-level code. I can do the translation later, when I find some time for it.

Reply:
Tests in `test.builtin` are testing Mathics3 Builtin functions.  Of course, any time one tests a higher-level function one  is testing all sorts of lower-level functions. So all of these tests are valid, it is just organized in my opinion in the wrong place.

Tests of the ExpressionPattern match function are good because they isolate all of the other kinds of stuff and we pinpoint the problem better. A MatchQ test could fail because of something totally unrelated to matching.

Another thing that these smaller unit tests do (which shouldn't be needed but in fact is used way too much) is to show developers how one can work with, test, and debug the smaller units. Ideally, if our APIs and modularity are done well,  the code needed to do this is not that much. Personally, I have found ways to improve modularity when i have been forced to these the smaller parts in isolation.

Reply:
> Yes, but in any case, at some point we can rewrite these tests as lower-level tests. I do not think that these tests actually test the behavior of `MatchQ` or `ReplaceAll`, but the low-level code. I can do the translation later, when I find some time for it.

My experience is that the rewrite tends to happen at a smaller rate than creating more code that is somewhat ill-conceived.  In fact, I am in the middle of a larger rewrite, for operators right now.

If you want to do the minimal thing, then move the MatchQ code to where MatchQ should be tested and the rule test to where rules are tested. At least, this honestly reflects what the test is doing. If later you want to rewrite, then at that time you can.

Reply:
And yes, in that case, the test should be under `test/eval/`

Reply:
> When we test the evaluation of an expression like `MatchQ[1+2 a, x_.+y_.*z_Symbol]`, many processes are involved: the main evaluation loop, the definitions of `Plus` and `Times` and their internal implementation, the pattern matching done in the evaluation loop, and eventually, the implementation of `matchQ` and the pattern matching mechanisms. If we want to know if the pattern matching works as it should, it would be better to use `mathics.eval.patterns.match`, which does not try to evaluate its arguments, but just implements the pattern matching. I think that this is the right level of testing for these tests.
> 

Ok.  If testing `Matcher.match()` in   `mathics.eval.patterns.match` is a good place to test against, then let's do that and that would be done in a `test.eval.pattern` module. 

I am just saying that if we're going to test temporarily using MatchQ, then `testing.core.patterns` is not the right testr module for this. 

As written, the test is extending MatchQ tests, even if the motivation for doing this is to verify what's going on at a lower level.  I leave to @aravindh-krishnamoorthy whether to rewrite using `Matcher.match()` inside `test.eval.pattern` or move the existing tests to `test.builtin.testing_expressions.expression_tests` and wherever Rules are testing.

> Lower levels involve testing whether the functions and methods used to evaluate matching do what they are supposed to. But at this point, there are functions for which I am still unable to say exactly what the expected behavior is, even after spending quite a long time trying to understand and document them.

Then i think it was a mistake to "optimize" what is not well understood nor,  as we have seen, not well covered in tests.

Reply:
Thanks for your understanding and patience. Sure, it if turns out to be too cumbersome, do as you say.

In fact, the messes in this code have been so great that often I find that I can't fix up a single mess in one go, but have to break things up into little pieces.  As you are learning, Pattern matching is one of those big messes. 

Down the line, we will use trie based machining akin to expreduce (and may try to copy its algorithm).  In preparation for that though, having more and better pattern-matching tests in advance will help greatly.